packages/metro-config/index.js

```diff
@@ -36,6 +39,53 @@ const INTERNAL_CALLSITES_REGEX = new RegExp(
   ].join('|'),
 );
 
+/**
+ * Resolves the root of an NPM or Yarn workspace, by traversing the file tree upwards from a `candidatePath` in the search for
+ * - a directory with a package.json
+ * - which has a `workspaces` array of strings
+ * - which (possibly via a glob) includes the project root
+ * @param {string} projectRoot Project root to find a workspace root for
+ * @param {string | undefined} candidatePath Current path to search from
+ * @returns Path of a workspace root or `undefined`
+ */
+function getWorkspaceRoot(projectRoot /*: string */, candidatePath /*: string */ = projectRoot) /*: string | void */ {
+  const packageJsonPath = path.resolve(candidatePath, 'package.json');
+  if (fs.existsSync(packageJsonPath)) {
+    try {
+      const { workspaces } = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
+      if (Array.isArray(workspaces)) {
+        // If one of the workspaces match the project root, this is the workspace root
+        // Note: While NPM workspaces doesn't currently support globs, Yarn does.
+        const matches = fastGlob.sync(workspaces, {
+          cwd: candidatePath,
+          onlyDirectories: true,
+          absolute: true,
+        });
+        if (matches.includes(projectRoot)) {
+          return candidatePath;
+        }
+      }
+    } catch (err) {
+      console.warn(`Failed reading or parsing ${packageJsonPath}:`, err);
```

Comment:
Hmm.. this is awkward, because if there is an error here for whatever reason (less chance if we check for permissions above), even if the user explicitly specifies `watchFolders`, they can't (in this design) avoid running this auto-detection, generating warnings that aren't relevant to them.

Ideally, this detection would only run if the user hasn't specified `watchFolders` themselves, but to do that I think we'd need to move it to the `@react-native/community-cli-plugin` - and tbh that might be a better place for it than amongst these static defaults. (WDYT @huntie ?)

Reply:
I think that sounds fair 👍🏻

[Expo does perform similar detection as part of its config defaults](https://github.com/expo/expo/blob/2a96b86e780a2cc6c4cf4d5b38849f108f0517b0/packages/%40expo/metro-config/src/ExpoMetroConfig.ts#L154), however that is (conceptually) downstream from this package. I'd rather locate config that is dependent on the _embedder_ (Community CLI, Expo), particularly if dynamic, outside this file. So yeah, into `@react-native/community-cli-plugin`.